### PR TITLE
fix: unwind stages with respect to commit threshold

### LINKED ISSUE #2274: Unwind with thresholds
## Problem

Currently, stage's unwind works as a single operation meaning it unwinds the whole requested range at a time. This is less than ideal, because it can hold a lot of database write operations in memory at a time, sometimes even leading to OOMs (e.g. unwind of a wide execution range).

## Potential Solution

Make most stages (Hashing, Execution, Merkle) able to unwind in chunks, committing intermediate changes to the database (akin to how `execute` function already does with `commit_threshold` parameter).

### PR DESCRIPTION
**Pull request by shekhirin**
Resolves https://github.com/paradigmxyz/reth/issues/2274

### CODE REVIEW

**Comment by shekhirin** [comment]
Not sure what would be the proper way to test it, see several approaches:
1. Custom `unwind` CLI command akin to `stage`
2. Rust test with multiple blocks, so we can check that unwind happens one by one if `commit_threshold = 1`. Currently, we insert only 2 blocks: genesis and another one, so it's not possible to test it https://github.com/paradigmxyz/reth/blob/174316936baab8839970273a7fc681993148a3bf/crates/stages/src/stages/execution.rs#L389-L393 But that way, we could also test commit thresholds for stage execution, that's nice.

**Comment by shekhirin** [comment]
regarding `reth stage` CLI for unwinding, I added a `--batch-size` argument that allows to set the commit threshold, but it fails now because we previously expected `commit_threshold == num_blocks`
```console
RUST_LOG=debug cargo run --bin reth -- stage --from 90000 --to 100000 --batch-size 10 execution
    Blocking waiting for file lock on package cache
    Blocking waiting for file lock on build directory
   Compiling reth v0.1.0 (/Users/shekhirin/Projects/reth/bin/reth)
    Finished dev [unoptimized + debuginfo] target(s) in 14.86s
     Running `target/debug/reth stage --from 90000 --to 100000 --batch-size 10 execution`
2023-05-02T05:40:14.779032Z  INFO reth 0.1.0 starting stage Execution
99990 100000 10
99990..=100000
2023-05-02T05:40:14.785799Z  INFO Unwind iteration finished to_block=90000 unwind_progress=99989 is_final_range=false
Error: Stage encountered a execution error in block 90002: EVM reported invalid transaction (0xff2465731e1247b42d0ad10456b28ba99825fd9321488b8e850520f9d3839449): Transaction(NonceTooLow { tx: 3, state: 27 }).

Caused by:
    EVM reported invalid transaction (0xff2465731e1247b42d0ad10456b28ba99825fd9321488b8e850520f9d3839449): Transaction(NonceTooLow { tx: 3, state: 27 })

Location:
    bin/reth/src/stage/mod.rs:188:17
```

So probably need to introduce the incremental execute/unwind logic to CLI too, as it works in the pipeline.

File: bin/reth/src/stage/mod.rs [hunk 1]
```diff
@@ -157,53 +154,40 @@ impl Command {
...
+                let unwind_output = stage.unwind(&mut tx, unwind).await?;
+                unwind.stage_progress = unwind_output.stage_progress;
             }
-            _ => {}
+        }
+
+        while let ExecOutput { stage_progress, done: false } = stage.execute(&mut tx, input).await?
+        {
+            input.stage_progress = Some(stage_progress)
         }
```
Comment (gakonst) [inline bin/reth/src/stage/mod.rs:191]: ah really nice!

File: crates/stages/src/pipeline/mod.rs [hunk 2]
```diff
@@ -300,6 +300,8 @@ where
 
                         self.listeners
                             .notify(PipelineEvent::Unwound { stage_id, result: unwind_output });
+
+                        tx.commit()?;
```
Comment (gakonst) [inline crates/stages/src/pipeline/mod.rs:304]: Makes sense - we move the commit from outside of the loop to inside given we want the unwind to respect the commit thresholds

**Review by gakonst — APPROVED** [review]
good with me - do we not need this unwind with thresholds logic in the other stages too?

**Comment by shekhirin** [comment]
> do we not need this unwind with thresholds logic in the other stages too?

yep, we need. Added to most of them, except:
- `Bodies`, `Headers`, `Merkle` – iiuc we don't have commit threshold logic in `execute` now. Do we want to have it in `unwind`?
- `Finish` – looks like we don't need it there

**Comment by mattsse** [comment]
can't really comment on this,
@rkrasiuk wdyt?

**Comment by rkrasiuk** [comment]
@shekhirin apologies for the delay, i'll try to review it tonight ❤️

File: crates/stages/src/stages/tx_lookup.rs [hunk 3]
```diff
@@ -175,7 +177,9 @@ impl<DB: Database> Stage<DB> for TransactionLookupStage {
             }
         }
 
-        Ok(UnwindOutput { stage_progress: input.unwind_to })
+        let unwind_progress = *range.start() - 1;
```
Comment (rkrasiuk) [inline crates/stages/src/stages/tx_lookup.rs:180]: might be worth making it one of the returned values of `unwind_block_range_with_threshold`

**Review by rkrasiuk — COMMENTED** [review]
lgtm overall, this should work. @gakonst do you think it's worth adding tests to the affected stages?

P.S. `stages::tx_lookup::tests::unwind_transaction_lookup` test is failing

File: crates/stages/src/stages/tx_lookup.rs (hunk 3 above)
Comment (shekhirin) [reply crates/stages/src/stages/tx_lookup.rs]: good point, fixed

**Review by gakonst — APPROVED** [review]
Good with me, thanks!

**Pull request merged**